Implement svd_lowrank and pca_lowrank for complex numbers

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127199
* __->__ #125580

We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

lezcano committed:
Implement svd_lowrank for complex numbers

We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

Fixes https://github.com/pytorch/pytorch/issues/122188

[ghstack-poisoned]

lezcano commented on test/test_linalg.py:
```diff
@@ -2436,26 +2436,13 @@ def run_subtest(actual_rank, matrix_size, batches, device, svd_lowrank, **option
 
-            # check if svd_lowrank produces same singular values as torch.svd
-            U, S, V = torch.svd(a)
-            self.assertEqual(s.shape, S.shape)
-            self.assertEqual(u.shape, U.shape)
-            self.assertEqual(v.shape, V.shape)
+            # check if svd_lowrank produces same singular values as linalg.svdvals
+            S = torch.linalg.svdvals(a)
             self.assertEqual(s, S)
 
-            if density == 1:
-                # actual_rank is known only for dense inputs
-                #
-                # check if pairs (u, U) and (v, V) span the same
-                # subspaces, respectively
-                u, s, v = u[..., :actual_rank], s[..., :actual_rank], v[..., :actual_rank]
-                U, S, V = U[..., :actual_rank], S[..., :actual_rank], V[..., :actual_rank]
-                self.assertEqual(u.mT.matmul(U).det().abs(), torch.ones(batches, device=device, dtype=dtype))
-                self.assertEqual(v.mT.matmul(V).det().abs(), torch.ones(batches, device=device, dtype=dtype))
```
All these are removed, as the fact that `U @ S @ V.mH = A` checked above already defines that the SVD is correct.

lezcano committed:
Update on "Implement svd_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano commented:
@pearu PTAL

pearu reviewed (changes requested):
I have a suggestion to update `svd/pca_lowrank` doc-strings for complex input cases. There is a note about removed check in test_svd_lowrank which I think should be restored. Also, should we run test_pca_lowrank with complex dtype input as well?

Otherwise, the simplified pieces of the lowrank svd algorithm look good to me! Thanks, @lezcano!

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano commented:
Addressed all the reviews.

pearu commented on torch/_lowrank.py:
```diff
@@ -144,48 +152,30 @@ def _svd_lowrank(
-        U = Q.matmul(U)
+
+    # Assume that A is tall
+    if m < n:
```
Just a note: previously, this corresponded to `if m < n or n > q` but now, the restriction `n > q` could be dropped because `mH` has became a view operation.

pearu reviewed (approved):
LGTM! Thanks, @lezcano!

lezcano commented:
@pytorchbot merge

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano commented:
@peterbell10 could you please approve?

pearu commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -17683,13 +17690,25 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
-                                    device_type='cuda')],
+                       DecorateInfo(toleranceOverride({torch.float32: tol(atol=4e-02, rtol=4e-02)
+                                                       torch.complex64: tol(atol=4e-02, rtol=4e-02)}),
+                                    'TestCommon', 'test_noncontiguous_samples'),],
```
Fix typo:
```suggestion
                                    'TestCommon', 'test_noncontiguous_samples'),
```

pearu reviewed (changes requested):
CI failures are real.

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

peterbell10 commented on torch/_lowrank.py:
```diff
@@ -57,27 +58,27 @@ def get_approximate_basis(
+        X = X - matmul(M, R)
+    Q = torch.linalg.qr(X).Q
+    for i in range(niter):
+        X = matmul(A.mH, Q)
```
No need to recompute `A.mH` each iteration? (I know it's just a view but still)

lezcano replied:
I mean, it's a view in a linalg operation that does lots of matmuls, a svd, etc. I prefer readability here.

peterbell10 replied:
Each tensor takes about 1 us to create, and for small sizes `svd_lowrank` takes on the order of 100s of us so we're talking about a 1% speedup for 1 line of code. I would take that trade personally, but maybe that's just me.

pearu replied:
Having
```python
A_H = A.mH
for i in range(niter):
    X = matmul(A_H, Q)
```
does not sacrifice readability, IMHO.

lezcano replied:
This function is to be run on very large matrices that `linalg.svd` cannot support due to OOM

pearu replied:
How this comments about OOM relates to moving a view operation out of the loop? AFAIK, there is no additional burden to memory allocations.

lezcano replied:
My point is that Peter's comment is moot. This function is to be called with huge matrices, so all this discussion is just about a premature optimisation.

pearu replied:
Got it!

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2129,20 +2135,9 @@ def sample_inputs_svd_lowrank(op_info, device, dtype, requires_grad=False, **kwa
-        # in the documentation
-        op_kwargs['M'] = make_tensor((*batch, m, n), dtype=dtype, device=device, requires_grad=requires_grad)
-        yield clone_sample(sample, **op_kwargs)
+        yield SampleInput(a, args=(b,), kwargs=dict(q=k, M=None), output_process_fn_grad=fn)
+        M = make_tensor((*batch, m, n), dtype=dtype, device=device, requires_grad=requires_grad)
+        yield SampleInput(a, args=(b,), kwargs=dict(q=k, M=M), output_process_fn_grad=fn)
```
```suggestion
        yield SampleInput(a, b, q=k, M=None).with_meta(output_process_fn_grad=fn)
        M = make_tensor((*batch, m, n), dtype=dtype, device=device, requires_grad=requires_grad)
        yield SampleInput(a, b, q=k, M=M).with_meta(output_process_fn_grad=fn)
```

lezcano replied:
I mean, sure.

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2100,27 +2100,33 @@ def error_inputs_T(self, device, has_ndims_error=False):
-    size = [1, 5, 10]
-
+    batches = [(), (2,)]
+    size = [3, 4]
```
I would feel more comfortable if reducing the number of sample inputs was left until another PR.

lezcano replied:
the tests were taking forever. It doesn't make sense to have more tests here than in most linalg functions, given how extremely slow it is.

peterbell10 replied:
I didn't say you couldn't change the test, just split it into another PR so I can see the tests still pass.

lezcano replied:
done

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano commented:
addressed

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano commented:
addressed

pearu commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2109,18 +2109,24 @@ def sample_inputs_singular_matrix_factors(op_info, device, dtype, requires_grad=
-            a = make_arg((*batch, m, k))
-            b = make_arg((*batch, n, k))
-            yield SampleInput(a, b, **kwargs)
+        k = 2
+        a = make_arg((*batch, m, k))
+        b = make_arg((*batch, n, k))
+        yield a, b
+    yield make_arg((3, 1)), make_arg((2, 1))
+    yield make_arg((1, 3, 2)), make_arg((1, 4, 2))
```
This change set belongs to https://github.com/pytorch/pytorch/pull/127199 . Here we should have
```suggestion
        for k in range(min(3, m, n)):
            a = make_arg((*batch, m, k))
            b = make_arg((*batch, n, k))
            yield a, b
```

pearu commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2109,18 +2109,24 @@ def sample_inputs_singular_matrix_factors(op_info, device, dtype, requires_grad=
-            a = make_arg((*batch, m, k))
-            b = make_arg((*batch, n, k))
-            yield SampleInput(a, b, **kwargs)
+        k = 2
```
This is likely the cause of current CI failures: `k` should be `min(2, m, n)`.

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

pearu commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2129,20 +2133,9 @@ def sample_inputs_svd_lowrank(op_info, device, dtype, requires_grad=False, **kwa
-        # in the documentation
-        op_kwargs['M'] = make_tensor((*batch, m, n), dtype=dtype, device=device, requires_grad=requires_grad)
-        yield clone_sample(sample, **op_kwargs)
+        yield SampleInput(a, b, q=k, M=None).with_metadata(output_process_fn_grad=fn)
+        M = make_tensor((*batch, m, n), dtype=dtype, device=device, requires_grad=requires_grad)
+        yield SampleInput(a, b, q=k, M=M).with_metadata(output_process_fn_grad=fn)
```
Notice that previously cloned samples were returned. It is likely that cloning is required here as well. Say:
```suggestion
        yield clone_sample(SampleInput(a, b, q=k, M=None)).with_metadata(output_process_fn_grad=fn)
        M = make_tensor((*batch, m, n), dtype=dtype, device=device, requires_grad=requires_grad)
        yield clone_sample(SampleInput(a, b, q=k, M=M)).with_metadata(output_process_fn_grad=fn)
```

pearu commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -17646,33 +17639,45 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
-                                    device_type='cuda')],
+                       DecorateInfo(toleranceOverride({torch.float32: tol(atol=1e-03, rtol=1e-03),
+                                                       torch.complex64: tol(atol=1e-03, rtol=1e-03)}),
+                                    'TestCommon', 'test_noncontiguous_samples'),
```
```suggestion
                                    'TestCommon', 'test_noncontiguous_samples', device_type='cuda'),
```
? Or is the tolerance override now required for cpu device as well?

lezcano replied:
it was also failing on CPU

peterbell10 replied:
Presumably only `complex64` is failing on CPU?

lezcano replied:
updated just for complex64. Let's hope CI passes, I had to bump tolerances on pca_lowrank on float32 for no apparent reason.

lezcano replied:
rolling back to the previous state, it also needs a bump on CPU now.

pearu commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -17683,13 +17688,25 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
-                                    device_type='cuda')],
+                       DecorateInfo(toleranceOverride({torch.float32: tol(atol=4e-02, rtol=4e-02),
+                                                       torch.complex64: tol(atol=4e-02, rtol=4e-02)}),
+                                    'TestCommon', 'test_noncontiguous_samples'),
```
Is this tolerance override now required for cpu device cases?

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano commented:
CI was passing, if now it's not, I'll revert the last changes and open issues for them.
Landing this PR is taking way too long just adjusting tolerances on different subsystems and different architectures, which adds absolutely nothing to the whole testing of the feature.

pearu commented:
> CI was passing, if now it's not, I'll revert the last changes and open issues for them.

I'd advice to fight to the end :)

> Landing this PR is taking way too long just adjusting tolerances on different subsystems and different architectures, which adds absolutely nothing to the whole testing of the feature.

I sympathize with this statement but unfortunately this is often a normal workflow when changing the state of existing tests leads to tolerance-related failures in certain tests just because these now use different sample inputs...

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano commented:
addressed

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -17693,13 +17703,25 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
-                                    device_type='cuda')],
+                       DecorateInfo(toleranceOverride({torch.float32: tol(atol=4e-02, rtol=4e-02),
+                                                       torch.complex64: tol(atol=4e-02, rtol=4e-02)}),
+                                    'TestCommon', 'test_noncontiguous_samples'),
```
Whats with the tolerance bumps for `float32`?

pearu replied:
Yeah, tolerance 0.04 looks unusually high for float32, it could indicate a bug.
@lezcano can you reproduce this locally? If so, forcing the samples to be contiguous (but valuewise the same as used in the noncontiguous testing) could provide an answer.

lezcano replied:
The error here is in the gradients, and here we are differentiating an approximation algorithm. We don't have any guarantees that differentiating the algorithm even converges to the derivative, even less that it converges in the same number of iterations!

peterbell10 replied:
Okay, but I thought these were supposed to be cosmetic cleanups that shouldn't effect numerics. So why are larger tolerances required?

pearu replied:
Locally, with
```
diff --git a/torch/testing/_internal/common_methods_invocations.py b/torch/testing/_internal/common_methods_invocations.py
index 21fb3b0561a..3d5ccb91ffe 100644
--- a/torch/testing/_internal/common_methods_invocations.py
+++ b/torch/testing/_internal/common_methods_invocations.py
@@ -17703,8 +17703,8 @@ op_db: List[OpInfo] = [
            supports_fwgrad_bwgrad=True,
            sample_inputs_func=sample_inputs_pca_lowrank,
            decorators=[skipCUDAIfNoCusolver, skipCPUIfNoLapack, with_tf32_off,
-                       DecorateInfo(toleranceOverride({torch.float32: tol(atol=4e-02, rtol=4e-02),
-                                                       torch.complex64: tol(atol=4e-02, rtol=4e-02)}),
+                       DecorateInfo(toleranceOverride({torch.float32: tol(atol=1e-06, rtol=1e-06),
+                                                       torch.complex64: tol(atol=4e-03, rtol=4e-03)}),
                                     'TestCommon', 'test_noncontiguous_samples'),
                        # FIXME This should be the following, but the toleranceOverride does not seem to do anything!
                        # DecorateInfo(toleranceOverride({torch.complex128: tol(atol=1e-04, rtol=1e-04)}),
```
`test_noncontiguous_samples_pca_lowrank` tests pass OK.

For smaller tolerances, these tests fail with:
```
Got different gradients for contiguous / non-contiguous inputs wrt input 1.
```
With a deterministic operation, such failures should basically never happen because the  operation result typically does not depend on data storage details.

However, `pca_lowrank` is not a deterministic operation as its algorithm uses `torch.randn` to find the initial approximation of a lowrank input. As a result, calling `pca_lowrank` several times with identical inputs will *not* produce identical results. This explains why different gradients are obtained when using contiguous or non-contiguous inputs (that valuewise are equal).

lezcano replied:
The only change here is the clone. I now regenerate all samples for `M is not None`. I think that was the issue

lezcano replied:
In an operation as complex as this, getting all the numbers pretty close, and one within 0.01, pretty much tells you that the gradient is alright. Even more so when the gradient is autogenerated and the forward pass is passing the tests...

pearu replied:
> The only change here is the clone. I now regenerate all samples for `M is not None`. I think that was the issue

So, it was https://github.com/pytorch/pytorch/pull/125580#discussion_r1616165627 after all. I cannot remember anymore after all these years why sample_close was required in the first place... could it be due to side-effects from autograd or a combination of things discussed above? Sounds like a fun task to find it out but OT here.

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Implement svd_lowrank and pca_lowrank for complex numbers"


We fix a number of bugs previously present in the complex
implementation.

We also heavily simplify the implementation, using, among
other things, that we now have conjugate views.

I saw there is a comment regarding how slow some checks on this
function are. As such, I removed quite a few of the combinations of inputs
to make the OpInfo lighter. I still left a couple relevant examples to not regress
coverage though.

Fixes https://github.com/pytorch/pytorch/issues/122188

cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

pearu reviewed (approved):
LGTM! Thanks, @lezcano!
